src/codegen/llvm/codegen_llvm.cc

```diff
@@ -289,6 +290,10 @@ llvm::Type* CodeGenLLVM::LLVMType(const Type& t) const {
       case 64: etype = llvm::Type::getDoubleTy(*ctx_); break;
       default: LOG(FATAL) << "do not support " << t;
     }
+  } else if (DatatypeRegistry::Global()->DatatypeRegistered(t.code())) {
```

Comment:
Will the data type still be preserved during codegen? Can we also simply change all allocate to rewrite to the storage type so this is not necessary?

Reply:
Hm, when I think about it, it shouldn't be preserved. I went back and handled Allocate in the datatype lowering pass.

Reply:
I was able to lower Allocates just fine. However, I similarly had to lower Loads. The way I did this was, when encountering a Load of a custom type, I replace it with a new Load with the same fields *except* the custom type is changed to the corresponding storage type. However, perhaps unsurprisingly, this doesn't work---I'm fairly certain it's because, if we have a simple program like:
```cast(cast(x, posit) + cast(y, posit), float32)```
it will compile to:
```
for each xi in x, cast xi to posit and store
for each yi in y, cast yi to posit and store
load each xi as the storage type (int16)
load each yi as the storage type (int16)
xi + yi and store
...
```
because we change the type of the loads, the type of the addition (and everything else downstream) is changed as well.

Do you have any suggestions on how to get around this? Do you still think it is possible to completely rewrite all types to storage types before we get to codegen?

Reply:
Ah, actually, I'm realizing part of the problem might be the *order* in which I'm lowering different expressions. Loads should get lowered last, seemingly. Once all of the operations (Casts and Adds, for now) on custom datatypes are lowered to Calls, then Loads can be lowered. I'm going to test this theory by having a separate lowering pass for Loads which runs after my current lowering pass.

Reply:
That seems to have fixed it.